Source/WebCore/html/HTMLElement.cpp

```diff
@@ -1449,6 +1461,9 @@ ExceptionOr<void> HTMLElement::hidePopoverInternal(FocusPreviousElement focusPre
 
     removeFromTopLayer();
 
+    if (popoverState() == PopoverState::Auto)
+        document().popoverStack().removeLast();
```

Comment:
There are some early returned between you adding to the stack and removing from the stack. I am worried we may end up leaving items in the stack and leak documents since the stack holds a strong ref to Nodes.

Reply:
I think the behaviour is the same as for m_topLayerElements in Document (add both in showPopover, remove both inn hidePopover) and we do not seem to hit the release assert for m_topLayerElements in Document::removedLastRef.

Reply:
You need to guarantee that any element added to the stack will be removed. Using some sort of scope variable would likely work.

Reply:
@cdumez @rwlbuis I wonder if it's actually worth caching `popoverStack()` (which should probably be called `autoPopoverList()` fwiw) or not, maybe that method should just build a vector on the go of the filtered top layer, after all, the top layer should be a small vector in the vast majority of cases.

Reply:
Top level and auto popover list are somewhat related:
https://html.spec.whatwg.org/#auto-popover-list

Anyway, I found removedFromAncestor takes guaranteed care of removing top layer for the Element but does not guarantee cleanup for the auto popover list, so I added some code to take care of that, hopefully that fixes the leak problem?